### ISSUE TITLE
Performance Hit - ROCm Version Changes

### ISSUE THREAD

**Issue by kiritigowda** [ROCmSoftwarePlatform/MIOpen#161]
With the ROCm version, 3.1.34 and above the OpenCL backend kernels are taking a performance hit.

## Test Case
`
./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
`

## ROCm 2.10 - MIOpenDriver CLI Time

```
root@simon:/opt/rocm/miopen/bin# ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.056871 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 3049, 237, 0.056871
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
root@simon:/opt/rocm/miopen/bin# ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.061760 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 2808, 218, 0.061760
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
root@simon:/opt/rocm/miopen/bin# ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.057191 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 3032, 235, 0.057191
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
root@simon:/opt/rocm/miopen/bin# ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.056640 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 3062, 238, 0.056640
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
root@simon:/opt/rocm/miopen/bin# ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm:0
GPU Kernel Time Forward Conv. Elapsed: 0.062364 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 2781, 216, 0.062364
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
```
## ROCm 3.1.34 - MIOpenDriver CLI Time
````
kiriti@simon:/opt/rocm/miopen/bin$ ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.100569 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 1724, 134, 0.100569
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
kiriti@simon:/opt/rocm/miopen/bin$ ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 1
GPU Kernel Time Forward Conv. Elapsed: 0.107289 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 1616, 125, 0.107289
Forward Convolution Verifies on CPU and GPU (1.64048e-08)
kiriti@simon:/opt/rocm/miopen/bin$ ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.102880 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 1686, 131, 0.102880
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
kiriti@simon:/opt/rocm/miopen/bin$ ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.101049 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 1716, 133, 0.101049
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
kiriti@simon:/opt/rocm/miopen/bin$ ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.100747 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 1721, 134, 0.100747
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
````

## ROCm 3.3 - MIOpenDriver CLI Timing
```
kiriti@simon:~$ MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.100907 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 1719, 133, 0.100907
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
kiriti@simon:~$ MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.101102 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 1715, 133, 0.101102
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
kiriti@simon:~$ MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver: conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 0
GPU Kernel Time Forward Conv. Elapsed: 0.102222 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 1696, 132, 0.102222
Forward Convolution Verifies on CPU and GPU (2.19793e-08)
```

## Test Cases to Replicate Issue
````
./bin/MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 64 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver pool -n 1 -c 64 -H 224 -W 224 -y 2 -x 2 -p 0 -q 0 -u 2 -v 2 -m max -t 1
./bin/MIOpenDriver conv -n 1 -c 64 -H 112 -W 112 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 128 -H 112 -W 112 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver pool -n 1 -c 128 -H 112 -W 112 -y 2 -x 2 -p 0 -q 0 -u 2 -v 2 -m max -t 1
./bin/MIOpenDriver conv -n 1 -c 128 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 256 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 256 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 256 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver pool -n 1 -c 256 -H 56 -W 56 -y 2 -x 2 -p 0 -q 0 -u 2 -v 2 -m max -t 1
./bin/MIOpenDriver conv -n 1 -c 256 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 512 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 512 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 512 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver pool -n 1 -c 512 -H 28 -W 28 -y 2 -x 2 -p 0 -q 0 -u 2 -v 2 -m max -t 1
./bin/MIOpenDriver conv -n 1 -c 512 -H 14 -W 14 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 512 -H 14 -W 14 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 512 -H 14 -W 14 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 512 -H 14 -W 14 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver pool -n 1 -c 512 -H 14 -W 14 -y 2 -x 2 -p 0 -q 0 -u 2 -v 2 -m max -t 1
./bin/MIOpenDriver conv -n 1 -c 512 -H 7 -W 7 -k 4096 -y 7 -x 7 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 4096 -H 1 -W 1 -k 4096 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
./bin/MIOpenDriver conv -n 1 -c 4096 -H 1 -W 1 -k 1000 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 
````

**Comment by daniellowell** [comment]
@alexandraBara Can you make sure these get into our config database.

**Comment by alexandraBara** [comment]
> @alexandraBara Can you make sure these get into our config database.
@daniellowell  you want these to be recurrent or just exist in our table?

**Comment by kiritigowda** [comment]
Any updates on this bug @daniellowell ? Thanks!

**Comment by atamazov** [comment]
@alexandraBara @daniellowell IIRC some time ago we excluded `-n 1` cases from tuning?

**Comment by alexandraBara** [comment]
> > @alexandraBara Can you make sure these get into our config database.
> 
> @daniellowell you want these to be recurrent or just exist in our table?

configs have been added.

**Comment by alexandraBara** [comment]
> @alexandraBara @daniellowell IIRC some time ago we excluded `-n 1` cases from tuning?
 we havent tuned batchsize 1 in a while i think. We are doing it in our current tuning.

**Comment by kiritigowda** [comment]
Any updates on this bug?

**Comment by kiritigowda** [comment]
@daniellowell is there a fix for this with ROCm 3.6 and above?

**Comment by kiritigowda** [comment]
with ROCm 3.7 and above, the performance looks to be back to normal, can you confirm the fix?

```
kiriti@ThreadRipperUbuntu:/opt/rocm/miopen/bin$ ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 3, Solution: 53/ConvBinWinogradRxSf2x3
GPU Kernel Time Forward Conv. Elapsed: 0.050453 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 3437, 267, 0.050453
Forward Convolution Verifies on CPU and GPU (2.44951e-08)
kiriti@ThreadRipperUbuntu:/opt/rocm/miopen/bin$ ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 3, Solution: 53/ConvBinWinogradRxSf2x3
GPU Kernel Time Forward Conv. Elapsed: 0.050293 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 3448, 268, 0.050293
Forward Convolution Verifies on CPU and GPU (2.44951e-08)
kiriti@ThreadRipperUbuntu:/opt/rocm/miopen/bin$ ./MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver conv -n 1 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpen Forward Conv. Algorithm: 3, Solution: 53/ConvBinWinogradRxSf2x3
GPU Kernel Time Forward Conv. Elapsed: 0.051307 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 1, 3, 224, 224, 3, 3, 64,  173408256, 609024, 12845056, 3380, 262, 0.051307
Forward Convolution Verifies on CPU and GPU (2.44951e-08)
```
